### ISSUE TITLE
[v.2.3.0] Release Tracker

### ISSUE THREAD

**Issue by atalman** [pytorch/pytorch#121760]
### 🐛 Describe the bug

We cut a [release branch](https://github.com/pytorch/pytorch/tree/release/2.3) for the 2.3.0 release.

Our plan from this point from this point is roughly:

* Phase 1 (until 4/1/24): work on finalizing the release branch
* Phase 2 (after 4/1/24): perform extended integration/stability/performance testing based on Release Candidate builds.

This issue is for tracking cherry-picks to the release branch.

## Cherry-Pick Criteria

**Phase 1 (until 4/1/24):**

Only low-risk changes may be cherry-picked from main:

1. Fixes to regressions against the most recent minor release (e.g. 2.2.x for this release; see [module: regression issue list](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22module%3A+regression%22+))
2. Critical fixes for: [silent correctness](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+correctness+%28silent%29%22), [backwards compatibility](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+bc-breaking%22+), [crashes](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+crash%22+), [deadlocks](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+deadlock%22+), (large) [memory leaks](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+memory+usage%22+)
3. Critical fixes to new features introduced in the most recent minor release (e.g. 2.2.x for this release)
4. Test/CI fixes
5. Documentation improvements
6. Compilation fixes or ifdefs required for different versions of the compilers or third-party libraries
7. Release branch specific changes (e.g. change version identifiers)

Any other change requires special dispensation from the release managers (currently  @atalman, @osalpekar, @huydhn, @malfet). If this applies to your change please write "Special Dispensation" in the "Criteria Category:" template below and explain.

**Phase 2 (after 4/1/24):**

Note that changes here require us to rebuild a Release Candidate and restart extended testing (likely delaying the release). Therefore, the only accepted changes are **Release-blocking** critical fixes for: [silent correctness](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+correctness+%28silent%29%22), [backwards compatibility](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+bc-breaking%22+), [crashes](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+crash%22+), [deadlocks](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+deadlock%22+), (large) [memory leaks](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+memory+usage%22+)

Changes will likely require a discussion with the larger release team over VC or Slack.

## Cherry-Pick Process

1. Ensure your PR has landed in master. This does not apply for release-branch specific changes (see Phase 1 criteria).
2. Create (but do not land) a PR against the [release branch](https://github.com/pytorch/pytorch/tree/release/2.3).
   <details>

    ```bash
    # Find the hash of the commit you want to cherry pick
    # (for example, abcdef12345)
    git log

    git fetch origin release/2.3
    git checkout release/2.3
    git cherry-pick -x abcdef12345

    # Submit a PR based against 'release/2.3' either:
    # via the GitHub UI
    git push my-fork

    # via the GitHub CLI
    gh pr create --base release/2.3
    ```

    </details>
3. Make a request below with the following format:

```
Link to landed trunk PR (if applicable):
* 

Link to release branch PR:
* 

Criteria Category:
* 
```

1. Someone from the release team will reply with approved / denied or ask for more information.
2. If approved, someone from the release team will merge your PR once the tests pass. **Do not land the release branch PR yourself.**

**NOTE: Our normal tools (ghstack / ghimport, etc.) do not work on the release branch.**

Please note HUD Link with branch CI status and link to the HUD to be provided here.
[HUD](https://hud.pytorch.org/hud/pytorch/pytorch/release%2F2.3)

### Versions

2.3.0

**Comment by snadampal** [comment]
Link to main PR (if applicable):
- https://github.com/pytorch/builder/pull/1716
- https://github.com/pytorch/builder/pull/1717

Link to 2.3 release branch PR:
- https://github.com/pytorch/builder/pull/1728
- https://github.com/pytorch/builder/pull/1730

Criteria Category:
The PRs improve the torch.compile performance on aarch64 linux systems substantially.

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* NA

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/121726
* https://github.com/pytorch/pytorch/pull/121765
* https://github.com/pytorch/pytorch/pull/121813
* https://github.com/pytorch/pytorch/pull/121851

Criteria Category:
* Release only required changes

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121556

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/121929

Criteria Category:
* Critical fix: Without this we get mismatches between the GLIBC and GLIBCXX ABI used by conda packages vs pytorch.

**Comment by NmomoN** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121892

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122065

Criteria Category:
* Fixes to new features being introduced in 2.3.0 release.

**Comment by shunting314** [comment]
Link to landed trunk PR (if applicable):
N/A

Link to release branch PR:
- https://github.com/pytorch/pytorch/pull/122139

Criteria Category:
- need this to make pytorch release work with the corresponding triton release

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* N/A

Link to release branch PR:
* https://github.com/pytorch/builder/pull/1751
* https://github.com/pytorch/pytorch/pull/122403

Criteria Category:
* need this to make pytorch release work with the corresponding triton release

**Comment by wz337** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121763

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122495

Criteria Category:
* Cherry-picking a revert. Bug fix.

**Comment by SS-JIA** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121598

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122634

Criteria Category:
* This does not quite fit into any of the cherry pick criterias, however, it implements features required for ExecuTorch alpha release. It is also low risk as it only impacts the PyTorch-Vulkan backend. cc: @huydhn

**Comment by guangy10** [comment]
Link to landed trunk PR (if applicable):

- #122524  

Link to release branch PR:

- #122524

Criteria Category:

A fix of CI regression required by ExecuTorch alpha release. cc: @huydhn

**Comment by JacobSzwejbka** [comment]
Link to landed trunk PR (if applicable):

https://github.com/pytorch/pytorch/pull/122246

Link to release branch PR:

https://github.com/pytorch/pytorch/pull/122655

Criteria Category:

A new pass that ExecuTorch uses to allow mutable state with custom ops. cc: @huydhn

**Comment by JacobSzwejbka** [comment]
Link to landed trunk PR (if applicable):

https://github.com/pytorch/pytorch/pull/121990

Link to release branch PR:

https://github.com/pytorch/pytorch/pull/122654

Criteria Category:

Patching auto functionalized which ExecuTorch relies on for exporting llama. cc: @huydhn

**Comment by JacobSzwejbka** [comment]
Link to landed trunk PR (if applicable):

https://github.com/pytorch/pytorch/pull/122683

Link to release branch PR:

https://github.com/pytorch/pytorch/pull/122721

Criteria Category:

Makes dead code elimination pass not trash the output of capture_pre_autograd_graph if it contains kv cache mutation.  cc: @huydhn

**Comment by jataylo** [comment]
Link to landed trunk PR (if applicable):
- Not applicable, release specific change.

Link to release branch PR:
- https://github.com/pytorch/pytorch/pull/122493

Criteria Category:
- Release specific change for ROCm to use release/2.3.x triton branch cc: @atalman

**Comment by guangy10** [comment]
Link to landed trunk PR (if applicable):
 - #113279

Link to release branch PR:
 - #122854

Criteria Category:

A critical fix for ExecuTorch. W/o this PR, PyTorch v2.3 will not be compatible with ExecuTorch, even the simples add module will fail.

**Comment by SS-JIA** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122096

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122861

Criteria Category:
* This does not quite fit into any of the cherry pick criterias, however, it implements fixes to PyTorch-Vulkan required for ExecuTorch alpha release. It is also low risk as it only impacts the PyTorch-Vulkan backend. cc: @huydhn

**Comment by htyu** [comment]
Link to landed trunk PR:
- https://github.com/pytorch/pytorch/pull/122852

Link to release branch PR:
- https://github.com/pytorch/pytorch/pull/122906

Criteria Category:
A critical fix for matmul performance on RocM.

**Comment by jithunnair-amd** [comment]
Link to landed trunk PR (if applicable):
https://github.com/pytorch/pytorch/pull/122857

Link to release branch PR:
https://github.com/pytorch/pytorch/pull/122967

Criteria Category:
A critical fix for Flash Attention performance on ROCm

**Comment by chunyuan-w** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122297

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123064

Criteria Category:
* Fixes to regressions of Inductor CPP wrapper dynamic shape inference against the 2.2 release: https://github.com/pytorch/pytorch/issues/122292

**Comment by wz337** [comment]
Link to landed trunk PR (if applicable):
- https://github.com/pytorch/pytorch/pull/122975

Link to release branch PR:
- https://github.com/pytorch/pytorch/pull/123073

Criteria Category:
- Critical fix for DeviceMesh to avoid recreating nccl communicators multiple times.

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122164

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122930

Criteria Category:
* Critical Fixes

**Comment by wanchaol** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122755

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123106

Criteria Category:
* Critical fixes for: silent correctness on tensor subclasses

**Comment by huydhn** [comment]
Link to landed trunk PR (if applicable):
- https://github.com/pytorch/pytorch/pull/121918

Link to release branch PR:
- https://github.com/pytorch/pytorch/pull/123197

Criteria Category:
- CI fix

**Comment by ajindal1** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122000

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123271

Criteria Category:
* Backward compatibility with ONNX Runtime Training.

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121381

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/122785

Criteria Category:
* Critical fixes https://github.com/pytorch/pytorch/issues/120899

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/121731

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123289

Criteria Category:
* Critical fixes https://github.com/pytorch/pytorch/issues/121583

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122899
* https://github.com/pytorch/pytorch/pull/122157

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123355
* https://github.com/pytorch/pytorch/pull/123354

Criteria Category:
* Numpy 2.0 pytorch/pytorch

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/builder/pull/1768
* https://github.com/pytorch/builder/pull/1769
* https://github.com/pytorch/builder/commit/dcc58c30b5e3803d45a823595f6a23262e487099
* https://github.com/pytorch/builder/pull/1747
* https://github.com/pytorch/builder/pull/1745

Link to release branch PR:
* https://github.com/pytorch/builder/pull/1770

Criteria Category:
* Numpy 2.0 support for pytorch/builder

**Comment by BowenBao** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/123205

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123363

Criteria Category:
* Critical fixes: we found beartype 0.18 released on 4/2/2024 would uncover a type annotation issue in current ONNX related code and raise error for every export call. This fix changes the behavior from raising an error to warning.

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/123125

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123387

Criteria Category:
* Critical fix https://github.com/pytorch/pytorch/issues/121258

**Comment by atalman** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/123417

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123442

Criteria Category:
* Critical fix https://github.com/pytorch/pytorch/issues/122705

**Comment by malfet** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/123234

Link to release branch PR:
* https://github.com/pytorch/pytorch/pull/123385

Criteria Category:
* Silent correctness fix/regression https://github.com/pytorch/pytorch/issues/122016

**Comment by cyyever** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/pytorch/pull/122874

Link to release branch PR:
*  N/A

Criteria Category:
*  Critical fix: A bug fix for torch.norm when called with value = 1.

**Comment by WeizhuoZhang-intel** [comment]
Link to landed trunk PR (if applicable):

- https://github.com/pytorch/pytorch/pull/122591
- https://github.com/pytorch/pytorch/pull/123544

Link to release branch PR:

- https://github.com/pytorch/pytorch/pull/123827

Criteria Category:
- Critical fix: Add skip model by device in Dynamo test for torchbench suite. Some of those models have OOM issue and other models did not support CPU currently, which will help to improve the CPU side passrate when using `runner.py`.

**Comment by huydhn** [comment]
Link to landed trunk PR (if applicable):
* https://github.com/pytorch/test-infra/pull/5095

Link to release branch PR:
*  https://github.com/pytorch/test-infra/pull/5096

Criteria Category:
*  CI fix to unblock ExecuTorch build

----
@huydhn merged
